pandas/_libs/parsers.pyx

```diff
@@ -1267,19 +1267,49 @@ cdef class TextReader:
             return self._string_convert(i, start, end, na_filter,
                                         na_hashset)
         elif is_categorical_dtype(dtype):
+            # TODO: I suspect that _categorical_convert could be
+            # optimized when dtype is an instance of CategoricalDtype
             codes, cats, na_count = _categorical_convert(
                 self.parser, i, start, end, na_filter,
                 na_hashset, self.c_encoding)
-            # sort categories and recode if necessary
             cats = Index(cats)
-            if not cats.is_monotonic_increasing:
+
+            # Determine if we should convert inferred string
+            # categories to a specialized type
+            if (isinstance(dtype, CategoricalDtype) and
+                    dtype.categories is not None):
+                if dtype.categories.is_numeric():
+                    # is ignore correct?
+                    cats = to_numeric(cats, errors='ignore')
+                elif dtype.categories.is_all_dates:
```

Comment:
I think this may leave open corner cases where strings don't map 1->1 with categories?  For example:
```
cats:
# DatetimeIndex(['2014-01-01'], dtype='datetime64[ns]', freq=None)

data:
# ['2014-01-01', '2014-01-01T00:00:00', '2014-01-01']
```

Reply:
Sorry, I don't follow. This passes:

```python
        dtype = {
            'b': CategoricalDtype([pd.Timestamp("2014")])
        }
        # Two representations of the same value
        data = "b\n2014-01-01\n2014-01-01T00:00:00"
        expected = pd.DataFrame({'b': Categorical([pd.Timestamp('2014')] * 2)})
        result = self.read_csv(StringIO(data), dtype=dtype)
        tm.assert_frame_equal(result, expected)
```

Reply:
Does `result['b']` not have duplicated categories?  Sorry, don't have it checked out locally, only guessing.

Reply:
No problem. It has multiple values, but the categories are unique.

```python
In [10]: pd.read_csv(StringIO(data), dtype=dtype).b.dtype
Out[10]: CategoricalDtype(categories=['2014-01-01'], ordered=False)
```

The `categories` passed to the `Categorical` constructor later on comes directly from `dtype.categories`, which is unique. The coercion is done on the *values* so it's OK if different string forms are coerced to the same value.